### ISSUE TITLE
[syncfusion_flutter_datagrid] BUG: RangeError when filtering and selecting

### ISSUE THREAD

**Issue by Alyxsqrd** [syncfusion/flutter-widgets#999]
Hello, hope you all are doing well! 😀 I noticed a RangeError exception is happening in my app when filtering then selecting any row.

## How to Reproduce

1. Create an SfDataGrid
2. Set `columnWidthCalculationRange: ColumnWidthCalculationRange.allRows`
3. Set `selectionMode: SelectionMode.single` OR `selectionMode: SelectionMode.singleDeselect`
4. Set `allowFiltering: true`
5. Apply a filter to one of the columns
6. Click/tap on any row in the filtered source
7. **RangeError exception will throw** ⚠️

## Possible Fix

The `endRowIndex` value is set to the number of rows BEFORE filtering, it's not being updated to the new row count AFTER filtering: https://github.com/syncfusion/flutter-widgets/blob/master/packages/syncfusion_flutter_datagrid/lib/src/datagrid_widget/runtime/column.dart#L646

## Stack Trace

```
════════ Exception caught by widgets library ═══════════════════════════════════
The following RangeError was thrown building ScrollViewWidget(dirty, state: _ScrollViewWidgetState#56065):
RangeError (index): Invalid value: Only valid value is 0: 1

The relevant error-causing widget was
SfDataGrid
When the exception was thrown, this was the stack
#0      List.[] (dart:core-patch/growable_array.dart:264:36)
#1      ColumnSizer._getCellWidth
#2      ColumnSizer._calculateCellWidth
#3      ColumnSizer._calculateAllCellsExceptHeaderWidth
#4      ColumnSizer._calculateAllCellsWidth
#5      ColumnSizer._getWidthBasedOnColumn
#6      ColumnSizer._sizerColumnWidth
#7      ColumnSizer._ensureColumnVisibility
#8      ColumnSizer._refresh
#9      refreshColumnSizer
#10     _ScrollViewWidgetState._updateColumnSizer
#11     _ScrollViewWidgetState.build
#12     StatefulElement.build
#13     ComponentElement.performRebuild
#14     StatefulElement.performRebuild
#15     Element.rebuild
#16     BuildOwner.buildScope
#17     WidgetsBinding.drawFrame
#18     RendererBinding._handlePersistentFrameCallback
#19     SchedulerBinding._invokeFrameCallback
#20     SchedulerBinding.handleDrawFrame
#21     SchedulerBinding._handleDrawFrame
#22     _invoke (dart:ui/hooks.dart:148:13)
#23     PlatformDispatcher._drawFrame (dart:ui/platform_dispatcher.dart:318:5)
#24     _drawFrame (dart:ui/hooks.dart:115:31)
════════════════════════════════════════════════════════════════════════════════

════════ Exception caught by widgets library ═══════════════════════════════════
RangeError (index): Invalid value: Only valid value is 0: 1
The relevant error-causing widget was
SfDataGrid
════════════════════════════════════════════════════════════════════════════════
```

**Comment by Alyxsqrd** [comment]
Seems like an easy fix. A temporary solution I made was to change the internal code of `lib/src/datagrid_widget/runtime/column.dart` on line 655 from `source.rows` to `source.effectiveRows`:

**BEFORE**

```dart
endRowIndex = dataGridConfiguration.source.rows.length - 1;
```

**AFTER**

```dart
endRowIndex = dataGridConfiguration.source.effectiveRows.length - 1;
```

Now it works perfectly! 👍

**Comment by Tamilarasan-Paranthaman** [comment]
Hi @Alyxsqrd,

I'm facing the same issue on my end. But we cannot proceed with the effective rows property to fetch the row index. We will proceed with the row's properties if you set the **ColumnWidthCalculationRange** to **allRows**. It would consider all the rows you provided instead of the visible rows. Because if you clear the filtering, the column width might not be set to the large string value, which was not filtered previously. Can you please confirm whether your requirements will be met?

**Comment by Alyxsqrd** [comment]
> Hi @Alyxsqrd,
> 
> 
> 
> I'm facing the same issue on my end. But we cannot proceed with the effective rows property to fetch the row index. We will proceed with the **rows** property if you set the **ColumnWidthCalculationRange** to **allRows**. It would consider all the rows you provided instead of the visible rows. Because if you clear the filtering, the column width might not be set to the large string value, which was not filtered previously. Can you please confirm whether your requirements will be met?

Yes, that should be fine if the RangeError problem can still be resolved. 👍🏽

**Comment by Tamilarasan-Paranthaman** [comment]
Hi @Alyxsqrd,

Thanks for your confirmation. We have considered this as a bug and logged a bug report in our feedback portal. We will fix the reported issue and include the changes in our upcoming weekly patch release, which is expected to be rolled out on February 28, 2023. We will let you know once it is published. We appreciate your patience and understanding until then.

Feedback Link: https://www.syncfusion.com/feedback/41136/the-rangeerror-exception-is-thrown-when-rebuilding-the-datagrid-after-applying-the 

Regards,
Tamilarasan

**Comment by Tamilarasan-Paranthaman** [comment]
@Alyxsqrd, the fix for the reported issue “The RangeError is thrown when rebuilding the DataGrid after applying the filtering” has been included in the latest DataGrid version **20.4.51**. [Update](https://pub.dev/packages/syncfusion_flutter_datagrid/changelog) your DataGrid to get the issue resolved.
